lib/SILOptimizer/Transforms/TempLValueOpt.cpp

```diff
@@ -0,0 +1,299 @@
+
+      if (inst == copyInst)
+        endOfLiferangeReached = true;
+    }
+    if (beginOfLiferange && !endOfLiferangeReached) {
+      // If the root address of the destination is within the liferange of the
+      // temporary, we cannot replace all uses of the temporary with the
+      // destination (it would break the def-use dominance rule).
+      if (inst == destRootInst)
+        return false;
+        
+      // Check if the destination is not accessed within the liferange of
+      // the temporary.
+      // This is unlikely, because the destination is initialized at the
+      // copyInst. But still, the destination could contain an initialized value
+      // which is destroyed before the copyInst.
+      if (AA->mayReadOrWriteMemory(inst, destination) &&
+          // Needed to treat init_existential_addr as not-writing projection.
+          projections.count(inst) == 0)
+        return false;
+    }
+  }
+  assert(endOfLiferangeReached);
+
+  // Move all projections of the destination address before the liferange of
+  // the temporary.
+  for (auto iter = beginOfLiferange->getIterator();
+       iter != copyInst->getIterator();) {
+    SILInstruction *inst = &*iter++;
+    if (projections.count(inst))
+      inst->moveBefore(beginOfLiferange);
+  }
+
+  if (!copyInst->isInitializationOfDest()) {
+    // Make sure the the destination is uninitialized before the liferange of
+    // the temporary.
+    SILBuilderWithScope builder(beginOfLiferange);
+    builder.createDestroyAddr(copyInst->getLoc(), destination);
+  }
+
+  // Replace all uses of the temporary with the destination address.
+  while (!temporary->use_empty()) {
+    Operand *use = *temporary->use_begin();
+    SILInstruction *user = use->getUser();
+    switch (user->getKind()) {
+      case SILInstructionKind::DeallocStackInst:
+        user->eraseFromParent();
+        break;
+      default:
+        use->set(destination);
+    }
+  }
+  temporary->eraseFromParent();
+  copyInst->eraseFromParent();
+  return true;
+}
+
+bool TempLValueOptPass::combineCopyAndDestroy(CopyAddrInst *copyInst) {
+  if (copyInst->isTakeOfSrc())
+    return false;
+
+  // Find a destroy_addr of the copy's source address.
+  DestroyAddrInst *destroy = nullptr;
+  for (Operand *srcUse : copyInst->getSrc()->getUses()) {
+    if ((destroy = dyn_cast<DestroyAddrInst>(srcUse->getUser())))
+      break;
+  }
+  SILBasicBlock *block = copyInst->getParent();
+  if (!destroy || destroy->getParent() != block)
+    return false;
+  assert(destroy->getOperand() == copyInst->getSrc());
+
+  // Check if the destroy_addr is after the copy_addr and if there are no
+  // memory writes between them.
+  for (auto iter = std::next(copyInst->getIterator());
+       iter != block->end(); ++iter) {
+    SILInstruction *inst = &*iter;
+    if (inst == destroy) {
+      copyInst->setIsTakeOfSrc(IsTake);
+      destroy->eraseFromParent();
+      return true;
+    }
+    if (inst->mayWriteToMemory())
```

Comment:
Here's where destroy hoisting does not check for reads. Maybe you didn't push the latest change yet.

Reply:
Yes, I forgot to change this. I pushed a new version with a fix